File: dart/lib/src/models/tag.dart
//
//  rerout
//  tag.dart
//
//  Official Dart SDK for the Rerout branded-link API.
//
//  Created by Ngonidzashe Mangudya on 20/05/2026.
//  Copyright (c) 2026 Codecraft Solutions
//  Licensed under the MIT License
//  https://codecraftsolutions.co.za
//

import 'package:meta/meta.dart';

/// A tag attached to a short link, as returned by the Rerout API.
@immutable
class Tag {
  /// Creates a [Tag].
  const Tag({required this.id, required this.name, required this.color});

  /// Parses a [Tag] from the API JSON envelope.
  factory Tag.fromJson(Map<String, dynamic> json) => Tag(
    id: json['id'] as String,
    name: json['name'] as String,
    color: json['color'] as String,
  );

  /// Stable tag identifier.
  final String id;

  /// Human-readable tag label.
  final String name;

  /// Display color — typically a hex string such as `#FF8800`.
  final String color;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Tag &&
          id == other.id &&
          name == other.name &&
          color == other.color;

  @override
  int get hashCode => Object.hash(id, name, color);

  @override
  String toString() => 'Tag(id: $id, name: $name, color: $color)';
}

File: php/src/Models/Tag.php
<?php

declare(strict_types=1);

namespace Rerout\Models;

use InvalidArgumentException;

/**
 * A tag attached to a link, as returned by the Rerout API.
 *
 * Fields mirror the server-side tag shape one-to-one so JSON is parsed without
 * transformation. Tags are read-only for API-key clients.
 */
final readonly class Tag
{
    /**
     * @param string $id    Tag identifier.
     * @param string $name  Human-readable tag name.
     * @param string $color Display colour (e.g. a hex string).
     */
    public function __construct(
        public string $id,
        public string $name,
        public string $color,
    ) {
    }

    /**
     * Parse a {@see Tag} from an API JSON object.
     *
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        return new self(
            id: self::requireString($data, 'id'),
            name: self::requireString($data, 'name'),
            color: self::requireString($data, 'color'),
        );
    }

    /**
     * @param array<string, mixed> $data
     */
    private static function requireString(array $data, string $key): string
    {
        $value = $data[$key] ?? null;
        if (!is_string($value)) {
            throw new InvalidArgumentException("Expected string field '{$key}' in Tag payload.");
        }

        return $value;
    }
}
